### PR TITLE
Add pyproject.toml & modern packaging to introduction.

### PR DESCRIPTION
**Pull request by hynek**
I have been surprised by the fact that pip-tools in fact does support pyproject.toml files, because the README is pretending like setup.py is the best way to go.

I've tried to somewhat rewrite it and point people to the new standard – let me know what you think.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -67,39 +68,62 @@ available. To compile from scratch, first delete the existing
...
+packages and applications and recommended for new projects. ``pip-compile``
+supports both installing your ``project.dependencies`` as well as your
+``project.optional-dependencies``. Thanks to the fact that this is an
+official standard, you can use ``pip-compile`` to pin your dependencies
+for projects based on modern standards-adhering packaging tools like
+`Hatch <https://hatch.pypa.io/>`_ or `flit <https://flit.pypa.io/>`_.
+
+Suppose you have a Django project that is packaged using ``Hatch``, and you
+want to pin it for production. You also want to pin your development tools
+into a separate pin file. You declare ``django`` as a dependency and create an
```
Comment (webknjaz) [inline README.rst:85]:
```suggestion
in a separate pin file. You declare ``django`` as a dependency and create an
```

File: README.rst [hunk 2]
```diff
@@ -67,39 +68,62 @@ available. To compile from scratch, first delete the existing
...
+    name = "my-cool-django-app"
+    version = "42"
+    dependencies = ["django"]
+    
+    [project.optional-dependencies]
+    dev = ["pytest"]
+
+You can produce your pin files as easily as:
 
 .. code-block:: bash
```
Comment (webknjaz) [inline README.rst:104]: While you're on it, snippets with prompts should use the console lexer:
```suggestion
.. code-block:: console
```

File: README.rst [hunk 3]
```diff
@@ -67,39 +68,62 @@ available. To compile from scratch, first delete the existing
...
 
-    $ pip-compile
-    #
```
Comment (webknjaz) [inline README.rst:80]: I think it's worth keeping the example output.

File: README.rst [hunk 4]
```diff
@@ -67,39 +68,62 @@ available. To compile from scratch, first delete the existing
...
+This is great for both pinning your applications, but also to keep the CI
+of your open-source Python package stable.
+
+Requirements from ``setup.py`` and ``setup.cfg``
+------------------------------------------------
 
-``pip-compile`` will produce your ``requirements.txt``, with all the Django
-dependencies (and all underlying dependencies) pinned.
+``pip-compile`` has also full support for ``setup.py``- and
+``setup.cfg``-based projects that drive setuptools.
```
Comment (webknjaz) [inline README.rst:116]: I think it's vice versa:
```suggestion
``setup.cfg``-based projects that use ``setuptools``.
```

**Review by webknjaz — COMMENTED** [review]
Got a few notes that may need addressing.

File: README.rst [hunk 5]
```diff
@@ -67,39 +68,105 @@ available. To compile from scratch, first delete the existing
...
 
-Without ``setup.py``
---------------------
+Just define your dependencies and extras as usual and run
+``pip-compile`` as above.
 
-If you don't use ``setup.py`` (`it's easy to write one`_), you can create a
-``requirements.in`` file to declare the Django dependency:
+``requirements.in``
+-------------------
```
Comment (atugushev) [inline README.rst:165]:
```suggestion
Requirements from ``requirements.in``
-------------------------------------
```

Minor: to be consistent with "Requirements from ``pyproject.toml``" section.

**Comment by hynek** [comment]
I have added a change to the help string of pip-compile.

**Review by webknjaz — APPROVED** [review]

**Pull request merged**